sydtest/src/Test/Syd/Diff.hs

```diff
@@ -0,0 +1,525 @@
+
+-- | Grouped 'String' diff
+--
+-- Like 'getStringDiff' but with entire strings instead of individual characters.
+getGroupedStringDiff :: String -> String -> [Diff String]
+getGroupedStringDiff actual expected = V.toList $ V.map (mapDiff V.toList) $ getGroupedVectorDiff (V.fromList actual) (V.fromList expected)
+
+-- | Diff two vectors
+--
+-- Prefer 'getGroupedVectorDiff' for performance reasons.
+getVectorDiff :: Eq a => Vector a -> Vector a -> Vector (Diff a)
+getVectorDiff = getVectorDiffBy (==)
+
+-- | Diff two vectors with different types using a custom equality operator
+--
+-- Prefer 'getGroupedVectorDiffBy' for performance reasons.
+getVectorDiffBy :: forall a b. (a -> b -> Bool) -> Vector a -> Vector b -> Vector (PolyDiff a b)
+getVectorDiffBy eq old new = computeDiffFromEditScript old new (getEditScriptBy eq old new)
+
+-- | Diff two vectors with grouped results
+getGroupedVectorDiff :: Eq a => Vector a -> Vector a -> Vector (Diff (Vector a))
+getGroupedVectorDiff = getGroupedVectorDiffBy (==)
+
+-- | Diff two vectors with grouped results using a custom equality operator
+getGroupedVectorDiffBy :: forall a b. (a -> b -> Bool) -> Vector a -> Vector b -> Vector (PolyDiff (Vector a) (Vector b))
+getGroupedVectorDiffBy eq old new = computeGroupedDiffFromEditScript old new (getEditScriptBy eq old new)
+
+-- | Compute the edit script to turn a given vector into the second given vector
+getEditScript :: forall a. Eq a => Vector a -> Vector a -> Vector Edit
+getEditScript = getEditScriptBy (==)
+
+-- | Compute the edit script to turn a given vector into the second given vector with a custom equality operator
+--
+-- From https://blog.robertelder.org/diff-algorithm/
+getEditScriptBy :: forall a b. (a -> b -> Bool) -> Vector a -> Vector b -> Vector Edit
+getEditScriptBy eq old new = V.fromList $ DList.toList $ runST $ go old new 0 0
+  where
+    go :: forall s. Vector a -> Vector b -> Int -> Int -> ST s (DList Edit)
+    go e f i j = do
+      -- N,M,L,Z = len(e),len(f),len(e)+len(f),2*min(len(e),len(f))+2
+      let upperN :: Int
+          upperN = V.length e
+
+      let upperM :: Int
+          upperM = V.length f
+
+      let upperL :: Int
+          upperL = upperN + upperM
+
+      let upperZ :: Int
+          upperZ = 2 * min upperN upperM + 2
+
+      -- if N > 0 and M > 0:
+      if upperN > 0 && upperM > 0
+        then do
+          -- w,g,p = N-M,[0]*Z,[0]*Z
+          let w :: Int
+              w = upperN - upperM
+
+          g <- MV.replicate upperZ 0 :: ST s (MVector s Int)
+
+          p <- MV.replicate upperZ 0 :: ST s (MVector s Int)
+
+          -- for h in range(0, (L//2+(L%2!=0))+1):
+          let hs :: [Int]
+              hs = [0 .. ((upperL `quot` 2) + (if odd upperL then 1 else 0))]
+
+          fmap fromJust $ -- We use fromJust because the algorithm is guaranteed to terminate
+            forUntilJust hs $ \h -> do
+              -- for r in range(0, 2):
+              forUntilJust [0, 1 :: Int] $ \r -> do
```

Comment:
It's hard to track these nested loops, e.g. figuring out whether the loop at line 192 is inside the loop at 178 or after it. The loop at 192 is inside the loop at 178, not after it, but I had to count the spaces in order to see that. This would be a lot easier to read if it was broken out into `where` clauses or several top-level function declarations.